queryable-spec.ts

```diff
@@ -0,0 +1,405 @@
+ * meaning it cannot be used for Bindings.
+ */
+export interface Stream<Q> extends EventEmitter {
+  read(): Q | null;
+}
+
+/**
+ * QueryOperationCost represents the cost of a given query operation.
+ */
+ interface QueryOperationCost {
+  /**
+   * An estimation of how many iterations over items are executed.
+   * This is used to determine the CPU cost.
+   */
+  iterations: number;
+  /**
+   * An estimation of how many items are stored in memory.
+   * This is used to determine the memory cost.
+   */
+  persistedItems: number;
+  /**
+   * An estimation of how many items block the stream.
+   * This is used to determine the time the stream is not progressing anymore.
+   */
+  blockingItems: number;
+  /**
+   * An estimation of the time to request items from sources.
+   * This estimation can be based on the `cardinality`, `pageSize`, and `requestTime` metadata entries.
+   * This is used to determine the I/O cost.
+   */
+  requestTime: number;
+  /**
+   * Custom properties
+   */
+  [key: string]: any;
+}
+
+/**
+ * QueryOperationOrder represents an ordering of the results of a given
+ * query operation. QueryOperationOrder objects can be returned by
+ * implementations of both the Filterable (quad orderings) and Queryable
+ * (bindings orderings) interfaces. Furthermore, QueryOperationObjects are
+ * used to represent both available orderings (i.e. orderings that may be
+ * requested by callers) and returned orderings (i.e. orderings followed
+ * by the returned iterators).
+ */
+interface QueryOperationOrder<T extends TermName | RDF.Variable> {
+  cost: QueryOperationCost;
+  terms: { term: T, direction: 'asc' | 'desc' }[];
+}
+
+interface QueryResultCardinality {
+  /**
+   * indicates the type of counting that was done, and MUST either be 
+   * "estimate" or "exact".
+   */
+  type: 'estimate' | 'exact';
+
+  /**
+   * Indicates an estimate of the number of quads in the stream if 
+   * type = "estimate", or the exact number of quads in the stream if 
+   * type = "exact".
+   */
+  value: number;
+}
+
+/**
+ * A QueryResultMetadata is an object that contains metadata about a certain
+ * query result.
+ */
+
+interface CardinalityMetadataOpts { 
+  cardinality: 'estimate' | 'exact'; 
+};
+
+interface AvailableOrdersMetadataOpts {
+  availableOrders: true;
+};
+
+interface BaseMetadataQuery<OrderItemsType extends TermName | RDF.Variable, AdditionalMetadataType extends {}> {
+  metadata<M extends CardinalityMetadataOpts | AvailableOrdersMetadataOpts>(opts?: M): Promise<
+    AdditionalMetadataType 
+      & (M extends CardinalityMetadataOpts ? { cardinality: QueryResultCardinality } : {})
+      & (M extends AvailableOrdersMetadataOpts ? { availableOrders: QueryOperationOrder<OrderItemsType>[] } : {})
+  >;
+}
+
+interface QueryExecuteOptions<OrderItemsType extends TermName | RDF.Variable> {
+  
+  /**
+   * TBD
+   */
+  order?: QueryOperationOrder<OrderItemsType>;
```

Comment:
Shouldn't this be 

```
order?: QueryOperationOrder<OrderItemsType>['terms'];
```

? 

The `cost` property here is irrelevant and one might want to request a specific ordering without necessarily checking which ones are available.

Reply:
It might be relevant if the user wants to check the current (=default) order's cost before execution. If that cost is too high, the user might still decide not to invoke `execute()`.

Reply:
Mh... I don't understand. This is the type of the `opts` argument of `execute()`, so the user has necessarily decided to invoke `execute()`?

Reply:
Oh, my bad, I thought this comment was about the `order` in the metadata.
So scratch my last comment, and replace that with "Yes, I agree" :-)